pyriemann/sampling.py

```diff
@@ -0,0 +1,325 @@
+    n_dim : int
+        dimensionality of the SPD matrices to be sampled
+    sigma : float
+        dispersion of the Riemannian Gaussian distribution
+    random_state : int, RandomState instance or None (default: 42)
+        pass an int for reproducible output
+
+    Returns
+    -------
+    samples : ndarray (n_samples, n_dim, n_dim)
+        samples of the Riemannian Gaussian distribution
+    """
+
+    samples_r = _sample_parameter_r(n_samples=n_samples,
+                                    n_dim=n_dim,
+                                    sigma=sigma,
+                                    random_state=random_state)
+    samples_U = _sample_parameter_U(n_samples=n_samples,
+                                    n_dim=n_dim,
+                                    random_state=random_state)
+
+    samples = np.zeros((n_samples, n_dim, n_dim))
+    for i in range(n_samples):
+        Ui = samples_U[i]
+        ri = samples_r[i]
+        samples[i] = Ui.T @ np.diag(np.exp(ri)) @ Ui
+        samples[i] = (samples[i] + samples[i].T) / 2.0  # ensure symmetry
+
+    return samples
+
+
+def sample_gaussian_spd(n_samples, mean, sigma, random_state=42):
+    """Sample a Riemannian Gaussian distribution
+
+    Sample SPD matrices from a Riemannian Gaussian distribution centered at
+    mean and with dispersion parametrized by sigma. This distribution has been
+    defined in Said et al. "Riemannian Gaussian Distributions on the space of
+    symmetric positive definite matrices" (2016) and generalizes the notion of
+    a Gaussian distribution to the space of SPD matrices. The sampling is based
+    on a spectral factorization of SPD matrices in terms of their eigenvectors
+    (U-parameters) and the log of the eigenvalues (r-parameters).
+
+    See https://arxiv.org/pdf/1507.01760.pdf for more details
+
+    Parameters
+    ----------
+    n_samples : int
+        how many samples to generate
+    mean : ndarray (n_dim, n_dim)
+        center of the Riemannian Gaussian distribution
+    sigma : float
+        dispersion of the Riemannian Gaussian distribution
+    random_state : int, RandomState instance or None (default: 42)
+        pass an int for reproducible output
+
+    Returns
+    -------
+    samples : ndarray (n_samples, n_dim, n_dim)
+        samples of the Riemannian Gaussian distribution
+    """
+
+    n_dim = mean.shape[0]
+    samples_centered = sample_gaussian_spd_centered(
+                        n_samples,
+                        n_dim=n_dim,
+                        sigma=sigma,
+                        random_state=random_state)
+
+    # apply the parallel transport to mean on each of the samples
+    samples = np.zeros((n_samples, n_dim, n_dim))
+    for i in range(n_samples):
+        samples[i] = sqrtm(mean) @ samples_centered[i] @ sqrtm(mean)
+        samples[i] = (samples[i] + samples[i].T) / 2.0  # ensure symmetry
+
+    return samples
+
+
+def generate_random_spd_matrix(n_dim, random_state=42):
```

Comment:
It could be useful to add a parameter `n_samples` in this function: `generate_random_spd_matrix(n_samples, n_dim, random_state=42)`

Reply:
My goal here was to have a very easy and simple way of creating one random SPD matrix, with no particular control over the statistical law it comes from. With that said, if someone wants to generate several random SPD matrices, he or she should prefer the `sample_gaussian_spd` function. Not sure whether this is a good idea, but I think that adding the `n_samples` argument makes the function a bit too cumbersome, since its output would have to be `(n_samples, n_dim, n_dim)`

Reply:
Ok.